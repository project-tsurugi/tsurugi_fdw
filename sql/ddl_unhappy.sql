/* Test case: unhappy path - Data Definition Language */
-- Test setup: DDL of the PostgreSQL
CREATE DATABASE contrib_regression2;
\c contrib_regression2

-- Test
CREATE TABLE tsurugifdw_regressiontest(id INT);
CREATE EXTENSION tsurugi_fdw;
\dx tsurugi_fdw
DROP TABLE tsurugifdw_regressiontest;

CREATE EXTENSION tsurugi_fdw;
\dx tsurugi_fdw
CREATE SERVER IF NOT EXISTS tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;

CREATE FOREIGN TABLE tsurugifdw_regressiontest(id INT) SERVER tsurugidb;
SELECT EXISTS (
  SELECT 1 FROM pg_class
    WHERE relname = 'tsurugifdw_regressiontest' AND relkind = 'f'
);

-- error (CREATE TABLE)
CREATE TABLE t0_create_table(c1 INTEGER NOT NULL);
SELECT * FROM t0_create_table;
CREATE TABLE IF NOT EXISTS mytable1 (id INT);
SELECT * FROM mytable1;
CREATE TABLE mytable2 (LIKE t0_create_table);
SELECT * FROM mytable2;
CREATE TABLE mytable3 (INHERITS t0_create_table);
SELECT * FROM mytable3;
CREATE TABLE mytable4 (id INT) PARTITION BY LIST (id);
SELECT * FROM mytable4;
CREATE TABLE mytable5 (id INT) USING heap;
SELECT * FROM mytable5;
CREATE TABLE mytable6 (id INT) WITH (fillfactor = 70);
SELECT * FROM mytable6;
CREATE UNLOGGED TABLE unlogged_table(id INT);
SELECT * FROM unlogged_table;
CREATE TEMPORARY TABLE temporary_table (id INT);
SELECT * FROM temporary_table;
CREATE LOCAL TEMPORARY TABLE temporary_table1 (id INT);
SELECT * FROM temporary_table1;
CREATE TEMP TABLE temp_table (id iNT);
SELECT * FROM temp_table;
CREATE LOCAL TEMP TABLE temp_table1 (id iNT);
SELECT * FROM temp_table1;

-- error (CREATE TABLE AS)
CREATE TABLE t1_create_table_as AS SELECT * FROM tsurugifdw_regressiontest;
SELECT * FROM t1_create_table_as;
SELECT * INTO t2_create_table_as FROM tsurugifdw_regressiontest;
SELECT * FROM t2_create_table_as;

DROP EXTENSION tsurugi_fdw CASCADE;

-- Success (CREATE TABLE)
CREATE TABLE t0_create_table(c1 INTEGER NOT NULL);
SELECT * FROM t0_create_table;
CREATE TABLE IF NOT EXISTS mytable1 (id INT);
SELECT * FROM mytable1;
CREATE TABLE mytable2 (LIKE t0_create_table);
SELECT * FROM mytable2;
CREATE TABLE mytable3 (INHERITS t0_create_table);
SELECT * FROM mytable3;
CREATE TABLE mytable4 (id INT) PARTITION BY LIST (id);
SELECT * FROM mytable4;
CREATE TABLE mytable5 (id INT) USING heap;
SELECT * FROM mytable5;
CREATE TABLE mytable6 (id INT) WITH (fillfactor = 70);
SELECT * FROM mytable6;
CREATE UNLOGGED TABLE unlogged_table(id INT);
SELECT * FROM unlogged_table;
CREATE TEMPORARY TABLE temporary_table (id INT);
SELECT * FROM temporary_table;
CREATE LOCAL TEMPORARY TABLE temporary_table1 (id INT);
SELECT * FROM temporary_table1;
CREATE TEMP TABLE temp_table (id iNT);
SELECT * FROM temp_table;
CREATE LOCAL TEMP TABLE temp_table1 (id iNT);
SELECT * FROM temp_table1;

-- Success (CREATE TABLE AS)
CREATE TABLE t1_create_table_as AS SELECT * FROM t0_create_table;
SELECT * FROM t1_create_table_as;
SELECT * INTO t2_create_table_as FROM t0_create_table;
SELECT * FROM t2_create_table_as;

-- Test teardown: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression2;
