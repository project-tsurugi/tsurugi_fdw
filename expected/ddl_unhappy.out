/* Test case: unhappy path - Data Definition Language */
-- EXTENSION
--- Test setup: DDL of the PostgreSQL
CREATE DATABASE contrib_regression_ddl;
\c contrib_regression_ddl
--- Test case: table exists before
CREATE TABLE tsurugifdw_regressiontest(id INT);
CREATE EXTENSION tsurugi_fdw;
ERROR:  tsurugi_fdw extension cannot be installed in the non-empty database. Please make sure there are no tables by using the \d command.
\dx tsurugi_fdw
     List of installed extensions
 Name | Version | Schema | Description 
------+---------+--------+-------------
(0 rows)

DROP TABLE tsurugifdw_regressiontest;
--- Test case: table does not exist
CREATE EXTENSION tsurugi_fdw;
\dx tsurugi_fdw
                   List of installed extensions
    Name     | Version | Schema |           Description            
-------------+---------+--------+----------------------------------
 tsurugi_fdw | 1.3.0   | public | foreign-data wrapper for tsurugi
(1 row)

--- Test case: DDL restriction - tsurugi_fdw is enabled
---- Test setup: DDL of the PostgreSQL
CREATE SERVER IF NOT EXISTS tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
CREATE FOREIGN TABLE tsurugifdw_regressiontest(id INT) SERVER tsurugidb;
SELECT EXISTS (
  SELECT 1 FROM pg_class
    WHERE relname = 'tsurugifdw_regressiontest' AND relkind = 'f'
);
 exists 
--------
 t
(1 row)

---- CREATE TABLE
CREATE TABLE t0_create_table(c1 INTEGER NOT NULL);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM t0_create_table;
ERROR:  relation "t0_create_table" does not exist
LINE 1: SELECT * FROM t0_create_table;
                      ^
CREATE TABLE IF NOT EXISTS mytable1 (id INT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable1;
ERROR:  relation "mytable1" does not exist
LINE 1: SELECT * FROM mytable1;
                      ^
CREATE TABLE mytable2 (LIKE t0_create_table);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable2;
ERROR:  relation "mytable2" does not exist
LINE 1: SELECT * FROM mytable2;
                      ^
CREATE TABLE mytable3 (INHERITS t0_create_table);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable3;
ERROR:  relation "mytable3" does not exist
LINE 1: SELECT * FROM mytable3;
                      ^
CREATE TABLE mytable4 (id INT) PARTITION BY LIST (id);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable4;
ERROR:  relation "mytable4" does not exist
LINE 1: SELECT * FROM mytable4;
                      ^
CREATE TABLE mytable5 (id INT) USING heap;
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable5;
ERROR:  relation "mytable5" does not exist
LINE 1: SELECT * FROM mytable5;
                      ^
CREATE TABLE mytable6 (id INT) WITH (fillfactor = 70);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM mytable6;
ERROR:  relation "mytable6" does not exist
LINE 1: SELECT * FROM mytable6;
                      ^
CREATE UNLOGGED TABLE unlogged_table(id INT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM unlogged_table;
ERROR:  relation "unlogged_table" does not exist
LINE 1: SELECT * FROM unlogged_table;
                      ^
CREATE TEMPORARY TABLE temporary_table (id INT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM temporary_table;
ERROR:  relation "temporary_table" does not exist
LINE 1: SELECT * FROM temporary_table;
                      ^
CREATE LOCAL TEMPORARY TABLE temporary_table1 (id INT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM temporary_table1;
ERROR:  relation "temporary_table1" does not exist
LINE 1: SELECT * FROM temporary_table1;
                      ^
CREATE TEMP TABLE temp_table (id iNT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM temp_table;
ERROR:  relation "temp_table" does not exist
LINE 1: SELECT * FROM temp_table;
                      ^
CREATE LOCAL TEMP TABLE temp_table1 (id iNT);
ERROR:  This database is for Tsurugi, so CREATE TABLE is not supported
SELECT * FROM temp_table1;
ERROR:  relation "temp_table1" does not exist
LINE 1: SELECT * FROM temp_table1;
                      ^
---- CREATE TABLE AS
CREATE TABLE t1_create_table_as AS SELECT * FROM tsurugifdw_regressiontest;
ERROR:  This database is for Tsurugi, so CREATE TABLE AS is not supported
SELECT * FROM t1_create_table_as;
ERROR:  relation "t1_create_table_as" does not exist
LINE 1: SELECT * FROM t1_create_table_as;
                      ^
SELECT * INTO t2_create_table_as FROM tsurugifdw_regressiontest;
ERROR:  This database is for Tsurugi, so CREATE TABLE AS is not supported
SELECT * FROM t2_create_table_as;
ERROR:  relation "t2_create_table_as" does not exist
LINE 1: SELECT * FROM t2_create_table_as;
                      ^
--- Test case: DDL restriction - tsurugi_fdw is disabled
---- Test setup: DDL of the PostgreSQL
DROP EXTENSION tsurugi_fdw CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to server tsurugidb
drop cascades to foreign table tsurugifdw_regressiontest
---- CREATE TABLE
CREATE TABLE t0_create_table(c1 INTEGER NOT NULL);
SELECT * FROM t0_create_table;
 c1 
----
(0 rows)

CREATE TABLE IF NOT EXISTS mytable1 (id INT);
SELECT * FROM mytable1;
 id 
----
(0 rows)

CREATE TABLE mytable2 (LIKE t0_create_table);
SELECT * FROM mytable2;
 c1 
----
(0 rows)

CREATE TABLE mytable3 (INHERITS t0_create_table);
SELECT * FROM mytable3;
 inherits 
----------
(0 rows)

CREATE TABLE mytable4 (id INT) PARTITION BY LIST (id);
SELECT * FROM mytable4;
 id 
----
(0 rows)

CREATE TABLE mytable5 (id INT) USING heap;
SELECT * FROM mytable5;
 id 
----
(0 rows)

CREATE TABLE mytable6 (id INT) WITH (fillfactor = 70);
SELECT * FROM mytable6;
 id 
----
(0 rows)

CREATE UNLOGGED TABLE unlogged_table(id INT);
SELECT * FROM unlogged_table;
 id 
----
(0 rows)

CREATE TEMPORARY TABLE temporary_table (id INT);
SELECT * FROM temporary_table;
 id 
----
(0 rows)

CREATE LOCAL TEMPORARY TABLE temporary_table1 (id INT);
SELECT * FROM temporary_table1;
 id 
----
(0 rows)

CREATE TEMP TABLE temp_table (id iNT);
SELECT * FROM temp_table;
 id 
----
(0 rows)

CREATE LOCAL TEMP TABLE temp_table1 (id iNT);
SELECT * FROM temp_table1;
 id 
----
(0 rows)

---- CREATE TABLE AS
CREATE TABLE t1_create_table_as AS SELECT * FROM t0_create_table;
SELECT * FROM t1_create_table_as;
 c1 
----
(0 rows)

SELECT * INTO t2_create_table_as FROM t0_create_table;
SELECT * FROM t2_create_table_as;
 c1 
----
(0 rows)

--- Test teardown: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression_ddl;
-- SERVER OPTIONS
--- Test setup: DDL of the Tsurugi
SELECT tg_execute_ddl('
  CREATE TABLE fdw_ddl_table (c INTEGER)
', 'tsurugidb');
 tg_execute_ddl 
----------------
 CREATE TABLE
(1 row)

--- Test setup: DDL of the PostgreSQL
CREATE DATABASE contrib_regression_ddl;
\c contrib_regression_ddl
CREATE EXTENSION tsurugi_fdw;
--- Test case: option name validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (invalid_option 'value');
ERROR:  invalid option "invalid_option"
--- Test case: required options validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream');
ERROR:  missing required option "address"
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1');
ERROR:  missing required option "port"
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', port '12345');
ERROR:  missing required option "address"
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
ALTER SERVER tsurugidb OPTIONS (endpoint 'stream');
ERROR:  missing required option "address"
ALTER SERVER tsurugidb OPTIONS (endpoint 'stream', address '127.0.0.1');
ERROR:  missing required option "port"
ALTER SERVER tsurugidb OPTIONS (endpoint 'stream', port '12345');
ERROR:  missing required option "address"
DROP SERVER tsurugidb;
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port '12345');
ALTER SERVER tsurugidb OPTIONS (DROP address);
ERROR:  missing required option "address"
ALTER SERVER tsurugidb OPTIONS (DROP port);
ERROR:  missing required option "port"
ALTER SERVER tsurugidb OPTIONS (DROP address, DROP port);
ERROR:  missing required option "address"
DROP SERVER tsurugidb;
--- Test case: endpoint option validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'invalid_value');
ERROR:  invalid value ("invalid_value") for option "endpoint"
DETAIL:  expected 'ipc' or 'stream'
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'IPC');
ERROR:  invalid value ("IPC") for option "endpoint"
DETAIL:  expected 'ipc' or 'stream'
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'STREAM');
ERROR:  invalid value ("STREAM") for option "endpoint"
DETAIL:  expected 'ipc' or 'stream'
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint ' ipc');
ERROR:  invalid value (" ipc") for option "endpoint"
DETAIL:  expected 'ipc' or 'stream'
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'ipc ');
ERROR:  invalid value ("ipc ") for option "endpoint"
DETAIL:  expected 'ipc' or 'stream'
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint '');
ERROR:  option "endpoint" requires a non-empty value
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint ' ');
ERROR:  option "endpoint" requires a non-empty value
--- Test case: dbname option validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'ipc', dbname '');
ERROR:  option "dbname" requires a non-empty value
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'ipc', dbname ' ');
ERROR:  option "dbname" requires a non-empty value
--- Test case: address option validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '', port '12345');
ERROR:  option "address" requires a non-empty value
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address ' ', port '12345');
ERROR:  option "address" requires a non-empty value
--- Test case: port option validation
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port '');
ERROR:  option "port" requires a non-empty value
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port ' ');
ERROR:  option "port" requires a non-empty value
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port '0');
ERROR:  "port" must be an integer value greater than zero: 0
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port 'abc');
ERROR:  "port" must be an integer value greater than zero: abc
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw
  OPTIONS (endpoint 'stream', address '127.0.0.1', port '-1');
ERROR:  "port" must be an integer value greater than zero: -1
--- Test case: dbname option value is incorrect
---- Test setup: DDL of the PostgreSQL
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
CREATE FOREIGN TABLE fdw_ddl_table (c integer) SERVER tsurugidb;
---- correct -> incorrect
ALTER SERVER tsurugidb OPTIONS (ADD dbname 'incorrect_dbname');
INSERT INTO fdw_ddl_table VALUES (1);
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
UPDATE fdw_ddl_table SET c = c + 10;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
DELETE FROM fdw_ddl_table;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
SELECT * FROM fdw_ddl_table ORDER BY c;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
---- incorrect -> correct
ALTER SERVER tsurugidb OPTIONS (DROP dbname);
INSERT INTO fdw_ddl_table VALUES (2);
SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
 2
(1 row)

UPDATE fdw_ddl_table SET c = c + 20;
SELECT * FROM fdw_ddl_table ORDER BY c;
 c  
----
 22
(1 row)

DELETE FROM fdw_ddl_table;
SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
(0 rows)

---- Test setup: PREPARE statements
PREPARE fdw_prepare_ins(integer) AS INSERT INTO fdw_ddl_table VALUES ($1);
PREPARE fdw_prepare_upd(integer) AS UPDATE fdw_ddl_table SET c = c + $1;
PREPARE fdw_prepare_del AS DELETE FROM fdw_ddl_table;
PREPARE fdw_prepare_sel AS SELECT * FROM fdw_ddl_table ORDER BY c;
---- correct -> incorrect (preparation)
ALTER SERVER tsurugidb OPTIONS (ADD dbname 'incorrect_dbname');
EXECUTE fdw_prepare_ins(1);
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
EXECUTE fdw_prepare_upd(10);
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
EXECUTE fdw_prepare_del;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
EXECUTE fdw_prepare_sel;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect_dbname
---- incorrect -> correct (preparation)
ALTER SERVER tsurugidb OPTIONS (SET dbname 'tsurugi');
-- EXECUTE fdw_prepare_ins(2);
-- EXECUTE fdw_prepare_sel;
-- EXECUTE fdw_prepare_upd(20);
-- EXECUTE fdw_prepare_sel;
-- EXECUTE fdw_prepare_del;
-- EXECUTE fdw_prepare_sel;
---- Test teardown: PREPARE statements
DEALLOCATE fdw_prepare_ins;
DEALLOCATE fdw_prepare_sel;
DEALLOCATE fdw_prepare_upd;
DEALLOCATE fdw_prepare_del;
---- Test setup: DDL of the PostgreSQL
DROP FOREIGN TABLE fdw_ddl_table;
DROP SERVER tsurugidb;
--- Operations during a transaction
---- Test re-setup: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression_ddl;
CREATE DATABASE contrib_regression_ddl;
\c contrib_regression_ddl
CREATE EXTENSION tsurugi_fdw;
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
CREATE FOREIGN TABLE fdw_ddl_table (c integer) SERVER tsurugidb;
---- Test case: Alter server (incorrect server) - Connected outside a transaction
SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
(0 rows)

BEGIN;
ALTER SERVER tsurugidb OPTIONS (dbname 'incorrect');
\des+
                                                List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |     FDW options      | Description 
-----------+----------+----------------------+-------------------+------+---------+----------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect') | 
(1 row)

SELECT * FROM fdw_ddl_table ORDER BY c;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect
END;
\des+
                                           List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version | FDW options | Description 
-----------+----------+----------------------+-------------------+------+---------+-------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         |             | 
(1 row)

SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
(0 rows)

---- Test case: Alter server (correct server) - Connected in a transaction
SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
(0 rows)

ALTER SERVER tsurugidb OPTIONS (dbname 'incorrect');
\des+
                                                List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |     FDW options      | Description 
-----------+----------+----------------------+-------------------+------+---------+----------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect') | 
(1 row)

BEGIN;
INSERT INTO fdw_ddl_table VALUES (1), (2);
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect
END;
BEGIN;
ALTER SERVER tsurugidb OPTIONS (SET dbname 'tsurugi');
-- INSERT INTO fdw_ddl_table VALUES (1), (2);
-- SELECT * FROM fdw_ddl_table ORDER BY c;
END;
---- Test re-setup: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression_ddl;
CREATE DATABASE contrib_regression_ddl;
\c contrib_regression_ddl
CREATE EXTENSION tsurugi_fdw;
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
CREATE FOREIGN TABLE fdw_ddl_table (c integer) SERVER tsurugidb;
---- Test case: ALTER SERVER during operations
BEGIN;
INSERT INTO fdw_ddl_table VALUES (3), (4);
ALTER SERVER tsurugidb OPTIONS (dbname 'incorrect-1');
\des+
                                                 List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |      FDW options       | Description 
-----------+----------+----------------------+-------------------+------+---------+------------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect-1') | 
(1 row)

UPDATE fdw_ddl_table SET c=c+10;
ALTER SERVER tsurugidb OPTIONS (SET dbname 'incorrect-2');
\des+
                                                 List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |      FDW options       | Description 
-----------+----------+----------------------+-------------------+------+---------+------------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect-2') | 
(1 row)

DELETE FROM fdw_ddl_table WHERE c=14;
ALTER SERVER tsurugidb OPTIONS (SET dbname 'incorrect-3');
\des+
                                                 List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |      FDW options       | Description 
-----------+----------+----------------------+-------------------+------+---------+------------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect-3') | 
(1 row)

SELECT * FROM fdw_ddl_table ORDER BY c;
 c  
----
 13
(1 row)

ALTER SERVER tsurugidb OPTIONS (SET dbname 'incorrect-4');
\des+
                                                 List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version |      FDW options       | Description 
-----------+----------+----------------------+-------------------+------+---------+------------------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         | (dbname 'incorrect-4') | 
(1 row)

END;
BEGIN;
SELECT * FROM fdw_ddl_table ORDER BY c;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect-4
END;
ALTER SERVER tsurugidb OPTIONS (DROP dbname);
\des+
                                           List of foreign servers
   Name    |  Owner   | Foreign-data wrapper | Access privileges | Type | Version | FDW options | Description 
-----------+----------+----------------------+-------------------+------+---------+-------------+-------------
 tsurugidb | postgres | tsurugi_fdw          |                   |      |         |             | 
(1 row)

BEGIN;
SELECT * FROM fdw_ddl_table ORDER BY c;
 c  
----
 13
(1 row)

DELETE FROM fdw_ddl_table;
END;
--- Test re-setup: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression_ddl;
CREATE DATABASE contrib_regression_ddl;
\c contrib_regression_ddl
CREATE EXTENSION tsurugi_fdw;
CREATE SERVER tsurugidb FOREIGN DATA WRAPPER tsurugi_fdw;
CREATE FOREIGN TABLE fdw_ddl_table (c integer) SERVER tsurugidb;
--- Test case: ALTER SERVER during operations
INSERT INTO fdw_ddl_table VALUES (1), (2);
ALTER SERVER tsurugidb OPTIONS (dbname 'incorrect');
SELECT * FROM fdw_ddl_table ORDER BY c;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect
ALTER SERVER tsurugidb OPTIONS (SET dbname 'tsurugi');
-- UPDATE fdw_ddl_table SET c=c+10;
ALTER SERVER tsurugidb OPTIONS (DROP dbname);
SELECT * FROM fdw_ddl_table ORDER BY c;
 c 
---
 1
 2
(2 rows)

ALTER SERVER tsurugidb OPTIONS (dbname 'incorrect');
DELETE FROM fdw_ddl_table;
ERROR:  Failed to execute remote SQL.
HINT:  Failed to make shared memory for Tsurugi. error: SERVER_FAILURE(10)
No detail message.
CONTEXT:  SQL query: incorrect
ALTER SERVER tsurugidb OPTIONS (SET dbname 'tsurugi');
-- DELETE FROM fdw_ddl_table;
-- SELECT * FROM fdw_ddl_table ORDER BY c;
--- Test teardown: DDL of the PostgreSQL
\c contrib_regression
DROP DATABASE contrib_regression_ddl;
--- Test teardown: DDL of the Tsurugi
SELECT tg_execute_ddl('DROP TABLE fdw_ddl_table', 'tsurugidb');
 tg_execute_ddl 
----------------
 DROP TABLE
(1 row)

